nav2_map_server/src/occ_grid_loader.cpp

```diff
@@ -27,100 +27,195 @@
   }
 
-  RCLCPP_DEBUG(node_->get_logger(), "resolution: %f", resolution_);
-  RCLCPP_DEBUG(node_->get_logger(), "origin[0]: %f", origin_[0]);
-  RCLCPP_DEBUG(node_->get_logger(), "origin[1]: %f", origin_[1]);
-  RCLCPP_DEBUG(node_->get_logger(), "origin[2]: %f", origin_[2]);
-  RCLCPP_DEBUG(node_->get_logger(), "free_thresh: %f", free_thresh_);
-  RCLCPP_DEBUG(node_->get_logger(), "occupied_thresh: %f", occupied_thresh_);
+  RCLCPP_DEBUG(node_->get_logger(), "resolution: %f", loadParameters.resolution);
+  RCLCPP_DEBUG(node_->get_logger(), "origin[0]: %f", loadParameters.origin[0]);
+  RCLCPP_DEBUG(node_->get_logger(), "origin[1]: %f", loadParameters.origin[1]);
+  RCLCPP_DEBUG(node_->get_logger(), "origin[2]: %f", loadParameters.origin[2]);
+  RCLCPP_DEBUG(node_->get_logger(), "free_thresh: %f", loadParameters.free_thresh);
+  RCLCPP_DEBUG(node_->get_logger(), "occupied_thresh: %f", loadParameters.occupied_thresh);
   RCLCPP_DEBUG(node_->get_logger(), "mode_str: %s", mode_str.c_str());
-  RCLCPP_DEBUG(node_->get_logger(), "mode: %d", mode_);
-  RCLCPP_DEBUG(node_->get_logger(), "negate: %d", negate_);
+  RCLCPP_DEBUG(node_->get_logger(), "mode: %d", loadParameters.mode);
+  RCLCPP_DEBUG(node_->get_logger(), "negate: %d", loadParameters.negate);
+
+  loadMapFromFile(map_filename, &loadParameters);
+
+  // Create a service callback handle
+  auto handle_occ_callback = [this](
+    const std::shared_ptr<rmw_request_id_t>/*request_header*/,
+    const std::shared_ptr<nav_msgs::srv::GetMap::Request>/*request*/,
+    std::shared_ptr<nav_msgs::srv::GetMap::Response> response) -> void {
+      RCLCPP_INFO(node_->get_logger(), "OccGridLoader: Handling map request");
+      response->map = *msg_;
+    };
+
+  // Create a service that provides the occupancy grid
+  occ_service_ = node_->create_service<nav_msgs::srv::GetMap>(service_name_, handle_occ_callback);
+
+  // Create a publisher using the QoS settings to emulate a ROS1 latched topic
+  rmw_qos_profile_t custom_qos_profile = rmw_qos_profile_default;
+  custom_qos_profile.depth = 1;
+  custom_qos_profile.durability = RMW_QOS_POLICY_DURABILITY_TRANSIENT_LOCAL;
+  custom_qos_profile.reliability = RMW_QOS_POLICY_RELIABILITY_RELIABLE;
+  occ_pub_ = node_->create_publisher<nav_msgs::msg::OccupancyGrid>(
+    topic_name_, custom_qos_profile);
+
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
+
+nav2_lifecycle::CallbackReturn
+OccGridLoader::on_activate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(node_->get_logger(), "OccGridLoader: Activating");
+
+  // Publish the map using the latched topic
+  occ_pub_->on_activate();
+  occ_pub_->publish(*msg_);
+
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
+}
+
+nav2_lifecycle::CallbackReturn
+OccGridLoader::on_deactivate(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(node_->get_logger(), "OccGridLoader: Deactivating");
+
+  occ_pub_->on_deactivate();
+
+  return nav2_lifecycle::CallbackReturn::SUCCESS;
```

Comment:
Should we be returning the value returned by the occ_pub_->on_deactivate() here?

Reply:
In general it looks like SUCCESS is always returned, shouldn't we be checking for errors and returning success or failure based on that?

Reply:
@mkhansen-intel  Since the lifecycle node catches exceptions and automatically transitions to on_error, I think the best way to do it is to throw an exception in any of these on_activate, on_deactivate, etc.

Reply:
Also, I think we should have some local data structures that can hold pointers to all of these sub-iterfaces, and automatically walk them through the state transitions. When we do that, the error detection and generation will be in one place.